controller/contour/cluster_role.go

```diff
@@ -0,0 +1,169 @@
+	verbGLW := []string{"get", "list", "watch"}
+
+	cfgMap := rbacv1.PolicyRule{
+		Verbs:     verbCGU,
+		APIGroups: groupAll,
+		Resources: []string{"configmaps"},
+	}
+	endPt := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupAll,
+		Resources: []string{"endpoints"},
+	}
+	secret := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupAll,
+		Resources: []string{"secrets"},
+	}
+	svc := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupAll,
+		Resources: []string{"services"},
+	}
+	crd := rbacv1.PolicyRule{
+		Verbs:     []string{"list"},
+		APIGroups: groupExt,
+		Resources: []string{"customresourcedefinitions"},
+	}
+	svcAPI := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupNet,
+		// TODO [danehans]: Update roles when v1alpha1 Service APIs are released.
+		Resources: []string{"gatewayclasses", "gateways", "httproutes", "tcproutes"},
+	}
+	ing := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupNet,
+		Resources: []string{"ingresses"},
+	}
+	ingStatus := rbacv1.PolicyRule{
+		Verbs:     verbCGU,
+		APIGroups: groupNet,
+		Resources: []string{"ingresses/status"},
+	}
+	cntr := rbacv1.PolicyRule{
+		Verbs:     verbGLW,
+		APIGroups: groupContour,
+		Resources: []string{"httpproxies", "tlscertificatedelegations", "extensionservices"},
+	}
+	cntrStatus := rbacv1.PolicyRule{
+		Verbs:     verbCGU,
+		APIGroups: groupContour,
+		Resources: []string{"httpproxies/status", "extensionservices/status"},
+	}
+
+	cr := oputil.NewClusterRole(name)
+	cr.Labels = map[string]string{
+		operatorv1alpha1.OwningContourNameLabel: contour.Name,
+		operatorv1alpha1.OwningContourNsLabel:   contour.Namespace,
+	}
+	cr.Rules = []rbacv1.PolicyRule{cfgMap, endPt, secret, svc, svcAPI, ing, ingStatus, cntr, cntrStatus, crd}
+	return cr
+}
+
+// currentClusterRole returns the current ClusterRole for the provided name.
+func (r *Reconciler) currentClusterRole(ctx context.Context, name string) (*rbacv1.ClusterRole, error) {
+	current := &rbacv1.ClusterRole{}
+	key := types.NamespacedName{Name: name}
+	err := r.Client.Get(ctx, key, current)
+	if err != nil {
+		return nil, err
+	}
+	return current, nil
+}
+
+// createClusterRole creates a ClusterRole resource for the provided cr.
+func (r *Reconciler) createClusterRole(ctx context.Context, cr *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	if err := r.Client.Create(ctx, cr); err != nil {
+		return nil, fmt.Errorf("failed to create cluster role %s: %w", cr.Name, err)
+	}
+	r.Log.Info("created cluster role", "name", cr.Name)
+
+	return cr, nil
+}
+
+// updateClusterRoleIfNeeded updates a ClusterRole resource if current does not match desired,
+// using contour to verify the existence of owner labels.
+func (r *Reconciler) updateClusterRoleIfNeeded(ctx context.Context, contour *operatorv1alpha1.Contour, current, desired *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
+	exist := doOwnerLabelsExist(current, contour)
+	if !exist {
```

Comment:
nit: I'd usually just write `if !doOwnerLabelsExist(...)` and drop the variable declaration since it's only used in one place. Not a big deal, just a little simpler IMHO. This applies to the other calls to `doOwnerLabelsExist` as well.